sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -388,3 +390,36 @@ object PhysicalWindow {
     case _ => None
   }
 }
+
+object ExtractSingleColumnNullAwareAntiJoin extends JoinSelectionHelper {
+
+  // SingleColumn NullAwareAntiJoin
+  // streamedSideKeys, buildSideKeys
+  // currently these two return Seq[Expression] should have only one element
+  private type ReturnType = (Seq[Expression], Seq[Expression])
+
+  /**
+   * See. [SPARK-32290]
+   * LeftAnti(condition: Or(EqualTo(a=b), IsNull(EqualTo(a=b)))
+   * will almost certainly be planned as a Broadcast Nested Loop join,
+   * which is very time consuming because it's an O(M*N) calculation.
+   * But if it's a single column case, and buildSide data is small enough,
+   * O(M*N) calculation could be optimized into O(M) using hash lookup instead of loop lookup.
+   */
+  def unapply(join: Join): Option[ReturnType] = join match {
+    case Join(left, right, LeftAnti,
+      Some(Or(EqualTo(leftAttr: AttributeReference, rightAttr: AttributeReference),
+        IsNull(EqualTo(tmpLeft: AttributeReference, tmpRight: AttributeReference)))), _)
+        if SQLConf.get.nullAwareAntiJoinOptimizeEnabled &&
+          leftAttr.semanticEquals(tmpLeft) && rightAttr.semanticEquals(tmpRight) &&
```

Comment:
Should this also refer to `canEvaluate` as done in `ExtractEquiJoinKeys`

Reply:
> Should this also refer to `canEvaluate` as done in `ExtractEquiJoinKeys`

I think semanticEquals is necessary. Because besides not in subquery can translate into LeftAnti

I can always write a SQL like

select * from a left anti join b on a.key = b.key OR isnull(a.key=b.keyB) which passed the canEvaluate of left or right plan. But it's not right. I think the pattern rule here is should be as Strict as we can.